jupyterhubutils/authenticator/lsstgithubauth.py

```diff
@@ -0,0 +1,170 @@
+        if denylist:
+            if not token:
+                self.log.warning("User does not have access token.")
+                userdict = None
+            else:
+                self.log.debug("Denylist `%s` found." % denylist)
+                denylist = denylist.split(',')
+                denied = yield self._check_denylist(userdict, denylist)
+            if denied:
+                self.log.warning("Rejecting user: denylisted")
+                userdict = None
+        self.lsst_mgr.uid = userdict["auth_state"]["github_user"]["id"]
+        return userdict
+
+    @gen.coroutine
+    def _set_groups_from_github_token(self, token):
+        self.log.debug("Acquiring list of user organizations.")
+        gh_org = yield self._get_github_user_organizations(token)
+        if not gh_org:
+            self.log.warning("Could not get list of user organizations.")
+        self.user_groups = list(gh_org.keys())
+        self.lsst_mgr.auth_mgr.group_map = gh_org
+        self.log.debug("Set user organizations to '{}'.".format(gh_org))
+
+    @gen.coroutine
+    def _check_github_denylist(self, userdict, denylist):
+        if ("auth_state" not in userdict or not userdict["auth_state"]):
+            self.log.warning("User doesn't have auth_state: rejecting.")
+            return True
+        ast = userdict["auth_state"]
+        if ("access_token" not in ast or not ast["access_token"]):
+            self.log.warning("User doesn't have access token: rejecting.")
+            return True
+        tok = ast["access_token"]
+        gh_org = yield self._get_github_user_organizations(tok)
+        if not gh_org:
+            self.log.warning("Could not get list of GH user orgs: rejecting.")
+            return True
+        deny = list(set(gh_org) & set(denylist))
+        if deny:
+            self.log.warning("User in denylist %s: rejecting." % str(deny))
+            return True
+        self.log.debug("User not in denylist %s" % str(denylist))
+        return False
+
+    @gen.coroutine
+    def _get_github_user_organizations(self, access_token):
+        # Requires 'read:org' token scope.
+        http_client = AsyncHTTPClient()
+        headers = github_api_headers(access_token)
+        gh_api = self.lsst_mgr.config.github_api
+        next_page = "https://{}/user/orgs".format(gh_api)
+        orgmap = {}
+        while next_page:
+            req = HTTPRequest(next_page, method="GET", headers=headers)
+            try:
+                resp = yield http_client.fetch(req)
+            except HTTPError as exc:
+                self.log.error("{} -> {}".format(next_page, exc))
+                return None
+            resp_json = json.loads(resp.body.decode('utf8', 'replace'))
+            next_page = next_page_from_links(resp)
+            for entry in resp_json:
+                # This could result in non-unique groups, if the first 32
+                #  characters of the group names are the same.
```

Comment:
because of NFS group limitations

Reply:
Removed.  16 NFS groups without a particular NFS server flag, no group name length limit (although user names are max 32 chars) that I can find.